[WASM] Avoid extraneous move for simple integer arithmetic in BBQ

ddegazio opened the pull request:
#### 7a2375d28da42fb41b734a98cda029478f96b611
<pre>
[WASM] Avoid extraneous move for simple integer arithmetic in BBQ
<a href="https://bugs.webkit.org/show_bug.cgi?id=272901">https://bugs.webkit.org/show_bug.cgi?id=272901</a>
<a href="https://rdar.apple.com/problem/126695575">rdar://problem/126695575</a>

Reviewed by Yusuke Suzuki.

Cleans up our code generation for integer add, subtract, bitwise ops,
and compare in the BBQ JIT. Specifically, we target the case where we
know we have one constant operand, and try and avoid a move.

 - For i32.add, i32.and, i32.or, and i32.xor, we are currently doing totally
   redundant work - three-operand nondestructive versions of all these opcodes
   with an immediate are supported on all MacroAssembler targets, so we should
   just use those.

 - For i32.sub, we currently do a weird trick involving adding a negative.
   The only real benefit of this is that for additions, we consider lowering as
   a LEA instruction on x86 - in this patch, we add that to subtractions in the
   MacroAssembler too.

 - For 64-bit versions of these operations, most MacroAssembler targets do not
   natively support 64-bit immediates. For those that do - ARM64 bitwise ops -
   we carve out a special case. Otherwise, we now check if the immediate is
   representable as a 32-bit immediate with no precision loss, and use the 32-bit
   immediate form instead if so.

* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
(JSC::MacroAssemblerX86Common::sub32):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::sub64):
* Source/JavaScriptCore/b3/B3Common.h:
(JSC::B3::isIdentical): Deleted.
(JSC::B3::isRepresentableAsImpl): Deleted.
(JSC::B3::isRepresentableAs): Deleted.
* Source/JavaScriptCore/b3/B3MemoryValue.cpp:
(JSC::B3::MemoryValue::isLegalOffsetImpl const):
* Source/JavaScriptCore/b3/B3ValueInlines.h:
(JSC::B3::Value::isRepresentableAs const):
* Source/JavaScriptCore/b3/air/AirArg.h:
(JSC::B3::Air::Arg::isRepresentableAs const):
(JSC::B3::Air::Arg::isRepresentableAs):
(JSC::B3::Air::Arg::isValidImmForm):
(JSC::B3::Air::Arg::isValidBitImmForm):
(JSC::B3::Air::Arg::isValidBitImm64Form):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Add):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Sub):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32And):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Xor):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Or):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Add):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Sub):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64And):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Xor):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Or):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCompareI64):
* Source/WTF/wtf/MathExtras.h:
(WTF::isIdentical):
(WTF::isRepresentableAsImpl):
(WTF::isRepresentableAs):

Canonical link: <a href="https://commits.webkit.org/277770@main">https://commits.webkit.org/277770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e4cbf1137d72801b37e5a05a0c571dde71324c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27716 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51464 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51192 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44570 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39659 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49087 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25396 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20772 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22875 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6561 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41801 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44814 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53098 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47995 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23551 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19872 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46971 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24816 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42063 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45890 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25621 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55490 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6916 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24539 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11412 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
How about adding add64 etc. with TrustedImm64 handling, which may use macro assembler scratch register if it is necessary?

hyjorc1 commented on Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
```diff
@@ -892,10 +892,20 @@ class MacroAssemblerX86Common : public AbstractMacroAssembler<Assembler> {
+        }
+
+        // Add a negative using lea
+        m_assembler.leal_mr(imm.m_value == INT32_MIN ? imm.m_value : -imm.m_value, left, dest);
```
Maybe we can use `WTF::negate`.

ddegazio replied:
Didn't know that existed! Definitely makes things more straightforward.

ddegazio replied:
Resolved in latest push.

hyjorc1 commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -964,8 +964,16 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+            move(a, dest);
+            sub64(imm, dest);
+        } else
+            m_assembler.leal_mr(-imm.m_value, left, dest);
```
typo, `left` -> `a`.

ddegazio replied:
Nice catch, will be fixed in next push.

Achierius commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -964,8 +964,16 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+            return;
+        }
+
+        if (UNLIKELY(imm.m_value == INT32_MIN)) {
```
Why not have a similar check for sub32?

ddegazio replied:
For `sub32`, we can just pass the `INT32_MIN` as a `TrustedImm32` - because it's a 32-bit operation, that's valid, because `-INT32_MIN == INT32_MIN` for 32-bit integers. For this function though, `-INT32_MIN` _is_ representable as a positive integer in the 64-bit value space, so it's not correct for us to truncate it to a 32-bit immediate in `lea`.

Achierius commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -2382,8 +2393,12 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addI64Xor(Value lhs, Value rhs, Value&
+#else
             m_jit.move(ImmHelpers::regLocation(lhsLocation, rhsLocation).asGPR(), resultLocation.asGPR());
             m_jit.xor64(TrustedImm64(ImmHelpers::imm(lhs, rhs).asI64()), resultLocation.asGPR());
+#endif
```
Idle thought, but given how this pattern shows up multiple times (esp. since it uses macros) it might be nice to extract this out into some sort of wrapper method which dispatches to either (move-then-arithmetic) or (just-arithmetic) depending on the target.

kmiller68 replied:
Yeah, we do this elsewhere in the MacroAssembler we should do it here too.

ddegazio replied:
Latest push ends up sidestepping this by moving the logic into the macro assembler.

hyjorc1 commented on Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
```diff
@@ -2310,8 +2310,12 @@ PartialResult WARN_UNUSED_RETURN BBQJIT::addI64Add(Value lhs, Value rhs, Value&
+            if (LIKELY(isRepresentableAs<int32_t>(ImmHelpers::imm(lhs, rhs).asI64())))
+                m_jit.add64(TrustedImm32(ImmHelpers::imm(lhs, rhs).asI64()), ImmHelpers::regLocation(lhsLocation, rhsLocation).asGPR(), resultLocation.asGPR());
+            else {
+                m_jit.move(ImmHelpers::regLocation(lhsLocation, rhsLocation).asGPR(), resultLocation.asGPR());
```
Since so far we only have 
```
void add64(TrustedImm32 imm, RegisterID src, RegisterID dest)
```
no
```
void add64(TrustedImm64 imm, RegisterID src, RegisterID dest)
```
for x86_64.

kmiller68 replied:
We should add `void add64(TrustedImm64 imm, RegisterID src, RegisterID dest)` in MacroAssembler since this will generate `i64.add reg, -1` as `move reg, dst; add64 -1, dst`. I think we historically didn't need such an overload before Wasm as it's not possible to get an int64 in JS.

ddegazio replied:
`isRepresentableAs` works okay for signed integers, so you would truly have to exceed the 32-bit integer range to need the move; but point taken, next push contains the addition of 64-bit variants of these instructions in the macro assembler.

Constellation reviewed (approved):
r=me

ddegazio committed:
[WASM] Avoid extraneous move for simple integer arithmetic in BBQ
https://bugs.webkit.org/show_bug.cgi?id=272901
rdar://problem/126695575

Reviewed by Yusuke Suzuki.

Cleans up our code generation for integer add, subtract, bitwise ops,
and compare in the BBQ JIT. Specifically, we target the case where we
know we have one constant operand, and try and avoid a move.

 - For i32.add, i32.and, i32.or, and i32.xor, we are currently doing totally
   redundant work - three-operand nondestructive versions of all these opcodes
   with an immediate are supported on all MacroAssembler targets, so we should
   just use those.

 - For i32.sub, we currently do a weird trick involving adding a negative.
   The only real benefit of this is that for additions, we consider lowering as
   a LEA instruction on x86 - in this patch, we add that to subtractions in the
   MacroAssembler too.

 - For 64-bit versions of these operations, most MacroAssembler targets do not
   natively support 64-bit immediates. For those that do - ARM64 bitwise ops -
   we carve out a special case. Otherwise, we now check if the immediate is
   representable as a 32-bit immediate with no precision loss, and use the 32-bit
   immediate form instead if so.

* Source/JavaScriptCore/assembler/MacroAssemblerX86Common.h:
(JSC::MacroAssemblerX86Common::sub32):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::sub64):
* Source/JavaScriptCore/b3/B3Common.h:
(JSC::B3::isIdentical): Deleted.
(JSC::B3::isRepresentableAsImpl): Deleted.
(JSC::B3::isRepresentableAs): Deleted.
* Source/JavaScriptCore/b3/B3MemoryValue.cpp:
(JSC::B3::MemoryValue::isLegalOffsetImpl const):
* Source/JavaScriptCore/b3/B3ValueInlines.h:
(JSC::B3::Value::isRepresentableAs const):
* Source/JavaScriptCore/b3/air/AirArg.h:
(JSC::B3::Air::Arg::isRepresentableAs const):
(JSC::B3::Air::Arg::isRepresentableAs):
(JSC::B3::Air::Arg::isValidImmForm):
(JSC::B3::Air::Arg::isValidBitImmForm):
(JSC::B3::Air::Arg::isValidBitImm64Form):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Add):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Sub):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32And):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Xor):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI32Or):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Add):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Sub):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64And):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Xor):
(JSC::Wasm::BBQJITImpl::BBQJIT::addI64Or):
(JSC::Wasm::BBQJITImpl::BBQJIT::emitCompareI64):
* Source/WTF/wtf/MathExtras.h:
(WTF::isIdentical):
(WTF::isRepresentableAsImpl):
(WTF::isRepresentableAs):

Canonical link: https://commits.webkit.org/277770@main
